tensorflow/core/kernels/cudnn_rnn_ops.cc

```diff
@@ -508,7 +508,7 @@ struct CudnnRnnModelShapes {
   bool IsCompatibleWith(const CudnnRnnModelShapes& rhs) const {
     return num_layers == rhs.num_layers && input_size == rhs.input_size &&
            num_units == rhs.num_units && dir_count == rhs.dir_count &&
-           cell_num_units == rhs.cell_num_units;
+           cell_num_units == rhs.cell_num_units && max_seq_length == rhs.max_seq_length;
```

Comment:
Let's add a `static_assert` on the size of the struct here so that we're reminded to update this method if we ever add new fields.

Reply:
@sanjoy Sure, something like `static_assert(sizeof(*this) == x);` ?

Reply:
That might sound a bit stupid, but I'm unsure whether you refer to `sizeof()` or something else more tensorflow specific: to the best of my knowledge sizeof(struct) has no guarantee to the the sum of its members size, so I'm unsure exactly what you have in mind here when you refer to "the size of the struct" :/

Reply:
I was talking about `sizeof`.  It is true that adding a new field will not necessarily increase the `sizeof` value, but it usually will and can be a good "quick and dirty" way to ensure that someone adding a new field remembers to update `IsCompatibleWith` and `CudnnRnnConfigHasher`.

Reply:
Hm, basic strict equality does not work for me at all :[
```
diff --git a/tensorflow/core/kernels/cudnn_rnn_ops.cc b/tensorflow/core/kernels/cudnn_rnn_ops.cc
index a7a80e25f4..503cc9cd6d 100644
--- a/tensorflow/core/kernels/cudnn_rnn_ops.cc
+++ b/tensorflow/core/kernels/cudnn_rnn_ops.cc
@@ -506,6 +506,20 @@ struct CudnnRnnModelShapes {
   TensorShape cell_state_shape;
   // At present only fields related to cached RnnDescriptor are concerned.
   bool IsCompatibleWith(const CudnnRnnModelShapes& rhs) const {
+    // Ensure the size of structure does not change, this will help
+    // people adding new field not forgetting about adding those fields to
+    // IsCompatibleWith()
+    static_assert(sizeof(struct CudnnRnnModelShapes) == sizeof(int /* num_layers */)
+                                 + sizeof(int /* input_size */)
+                                + sizeof(int /* num_units */)
+                                + sizeof(int /* dir_count */)
+                                + sizeof(int /* max_seq_length */)
+                                + sizeof(int /* batch_size */)
+                                + sizeof(int /* cell_num_units */)
+                                + sizeof(TensorShape /* input_shape */)
+                                + sizeof(TensorShape /* output_shape */)
+                                + sizeof(TensorShape /* hidden_state_shape */)
+                                + sizeof(TensorShape /* cell_state_shape */), "check struct CudnnRnnModelShapes members");
     return num_layers == rhs.num_layers && input_size == rhs.input_size &&
            num_units == rhs.num_units && dir_count == rhs.dir_count &&
            cell_num_units == rhs.cell_num_units && max_seq_length == rhs.max_seq_length;
```

Reply:
Greater or equal works, but I don't think it's really a useful check here :/

Reply:
FTR:
```
sizeof(struct CudnnRnnModelShapes) == 128
sizeof(int) * 7 == 28
sizeof(TensorShape) * 4 == 96
```

We could test for `sizeof(struct CudnnRnnModelShapes) >= 124 && sizeof(struct CudnnRnnModelShapes) <= 128`, but I feel it might blow easily :/

Reply:
Like, the static assert still works if I add a `char` or even an `int`, since the compiler can fit in the new entry within the padding it seems to have added. If you have a better idea, I'd be happy to implement it.

Reply:
Ok, probably then just a comment under `int cell_num_units = 0;` stating that if someone adds a new field they should also update the equality and hash functions?

Reply:
Done :)